File: tree/129.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sumNumbers(root *TreeNode) int {
    var ret int
    var cur []int
    dfs(root, &ret, cur)
    return ret
}

func dfs(root *TreeNode, ret *int, cur []int) {
    if root == nil {
        return
    }
    if root.Left == nil && root.Right == nil {
        *ret += sum(append(cur, root.Val))
    }else {
        dfs(root.Left, ret, append(cur, root.Val))
        dfs(root.Right, ret, append(cur, root.Val))
    }
}


func sum(numList []int) int {
    var sum int
    for i := 0; i < len(numList); i++{
        sum += numList[i] * int(math.Pow10(len(numList)-i-1))
    }
    
    return sum
}

File: tree/129.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        int sum = 0;
        recur(root, sum, 0);
        return sum;
    }
private:
    void recur(TreeNode* root, int& sum, int cur) {
        if (!root->left && !root->right) {
            sum += cur*10 + root->val;
            return;
        }
        if (root->left)
            recur(root->left, sum, cur*10+root->val);
        if (root->right)
            recur(root->right, sum, cur*10+root->val);
    }
};
